hadoop-tools/hadoop-aws/src/main/java/org/apache/hadoop/fs/s3a/s3guard/DynamoDBMetadataStoreTableHandler.java

```diff
@@ -0,0 +1,651 @@
+    final Item versionMarkerItem = getVersionMarkerItem();
+    final Item versionMarkerFromTag =
+        getVersionMarkerFromTags(table, amazonDynamoDB);
+
+    LOG.debug("versionMarkerItem: {};  versionMarkerFromTag: {}",
+        versionMarkerItem, versionMarkerFromTag);
+
+    if (versionMarkerItem == null && versionMarkerFromTag == null) {
+      if (!isEmptyTable(tableName, amazonDynamoDB)) {
+        LOG.error("Table is not empty but missing the version maker. Failing.");
+        throw new IOException(E_NO_VERSION_MARKER_AND_NOT_EMPTY
+            + " Table: " + tableName);
+      }
+
+      LOG.info("Table {} contains no version marker item or tag. " +
+              "The table is empty, so the version marker will be added " +
+              "as TAG and ITEM.", tableName);
+
+      tagTableWithVersionMarker();
+      putVersionMarkerToTable();
+    }
+
+    if (versionMarkerItem == null && versionMarkerFromTag != null) {
+      final int tagVersionMarker =
+          extractVersionFromMarker(versionMarkerFromTag);
+      throwExceptionOnVersionMismatch(tagVersionMarker, tableName,
+          E_INCOMPATIBLE_TAG_VERSION);
+
+      LOG.info("Table {} contains no version marker ITEM but contains " +
+              "compatible version marker TAG. Restoring the version marker " +
+              "item from tag.", tableName);
+
+      putVersionMarkerToTable();
+    }
+
+    if (versionMarkerItem != null && versionMarkerFromTag == null) {
+      final int itemVersionMarker =
+          extractVersionFromMarker(versionMarkerItem);
+      throwExceptionOnVersionMismatch(itemVersionMarker, tableName,
+          E_INCOMPATIBLE_ITEM_VERSION);
+
+      LOG.info("Table {} contains no version marker TAG but contains " +
+          "compatible version marker ITEM. Restoring the version marker " +
+          "item from item.", tableName);
+
+      tagTableWithVersionMarker();
+    }
+
+    if (versionMarkerItem != null && versionMarkerFromTag != null) {
+      final int tagVersionMarker =
+          extractVersionFromMarker(versionMarkerFromTag);
+      final int itemVersionMarker =
+          extractVersionFromMarker(versionMarkerItem);
+
+      throwExceptionOnVersionMismatch(tagVersionMarker, tableName,
+          E_INCOMPATIBLE_TAG_VERSION);
+      throwExceptionOnVersionMismatch(itemVersionMarker, tableName,
+          E_INCOMPATIBLE_ITEM_VERSION);
+
+      LOG.info("Table {} contains correct version marker TAG and ITEM.",
+          tableName);
+    }
+  }
+
+  private static boolean isEmptyTable(String tableName, AmazonDynamoDB aadb) {
+    final ScanRequest req = new ScanRequest().withTableName(
+        tableName).withLimit(1);
+    final ScanResult result = aadb.scan(req);
+    return result.getCount() == 0;
+  }
+
+  private static void throwExceptionOnVersionMismatch(int actual,
+      String tableName,
+      String exMsg) throws IOException {
+
+    if (VERSION != actual) {
+      throw new IOException(exMsg + " Table " + tableName
```

Comment:
or a PathIOException with the table as a Path? I'm never a fan of base IOEs

Reply:
I'd stay with PathIOException. But I like the idea of creating some specialized exception class for this.